experiments: Add focus-metavariable operator (#487)
returntocorp/semgrep-docs: docs/experiments/focus-metavariable.md

---
slug: focus-metavariable
append_help_link: true

> description: "focus-metavariable allows Semgrep users to narrow down the code matched by a rule, to focus on what it really matters."

Author: Maybe "The `focus-metavariable` operator ..."? Without "operator" it doesn't read well.

> description: "focus-metavariable allows Semgrep users to narrow down the code matched by a rule, to focus on what it really matters."

Reviewer: Yes, that is a good idea actually. Updating suggestion. @IagoAbal 

We should actually always define what we are talking about... if something is a flag, option, configuration variable, whatever... great you mention this.

---

# `focus-metavariable`

> The experimental `focus-metavariable` operator puts the focus, or "zooms in", on the code region matched by a metavariable. For example, if we wanted to find all functions arguments annotated with the type `bad` we could write this pattern:

Reviewer: Do not use quotation marks as emphasis. https://www.notion.so/r2cdev/r2c-technical-communication-style-guide-4d1104e78927460aa8c7624aedd6561f#b9b9ec08cbdf40e7ab5b79a0fa95e98b

> The experimental `focus-metavariable` operator puts the focus, or "zooms in", on the code region matched by a metavariable. For example, if we wanted to find all functions arguments annotated with the type `bad` we could write this pattern:

Reviewer: Thank you @IagoAbal. Updated suggestion.

```yaml
pattern: |
  def $FUNC(..., $ARG : bad, ...):
    ...
```

> This works but it matches the entire function definition. Sometimes that is not desirable, for instance, if we are using [Semgrep App](https://semgrep.dev/login) and we have triaged a finding generated by this pattern, the same finding may show up again as new if we make any change to the definition of the function!

Reviewer: ```suggestion
This works but it matches the entire function definition. Sometimes, this is not desirable, for example, if you are using [Semgrep App](https://semgrep.dev/login) and you have triaged a finding generated by this pattern, the same finding may show up again as new if you make any change to the definition of the function!
```

Reviewer: Check the use of [_we_ section in our style guide](https://www.notion.so/r2cdev/r2c-technical-communication-style-guide-4d1104e78927460aa8c7624aedd6561f#39871bd6aca04e37bac6dc2e70dd2baf). :-)

> Using `focus-metavariable` we can tell Semgrep that we are only interested in the code matched by a particular metavariable, in our case, `$ARG`:

Reviewer: ```suggestion
To specify that you are only interested in the code matched by a particular metavariable, in our case `$ARG`, use `focus-metavariable`.
```

Reviewer: (Just explanation of the suggestion above.) First, describe the goal (to do/achieve this and that), utilize this.

This makes the reading a little bit easier and it is more engaging for the reader.

<iframe src="https://semgrep.dev/embed/editor?snippet=4kk8" border="0" frameBorder="0" width="100%" height="435"></iframe>

> Note that `focus-metavariable: $ARG` is not the same as `pattern: $ARG`! Using `pattern: $ARG` will find all the uses of the parameter `x` which is not what we want! (Note that `pattern: $ARG` does not match the formal parameter declaration, because in this context `$ARG` only matches expressions.)

Reviewer: ```suggestion
Note that `focus-metavariable: $ARG` is not the same as `pattern: $ARG`! Using `pattern: $ARG` finds all the uses of the parameter `x` which is not what we want! (Note that `pattern: $ARG` does not match the formal parameter declaration, because in this context `$ARG` only matches expressions.)
```

<iframe src="https://semgrep.dev/embed/editor?snippet=PPPe" border="0" frameBorder="0" width="100%" height="435"></iframe>

> In short, `focus-metavariable: $X` is not a pattern in itself, it does not perform any matching, it just focuses the matching on the code already bound to `$X` by other patterns. Whereas `pattern: $X` will match `$X` against your code (and in this context, `$X` only matches expressions)!

Reviewer: ```suggestion
In short, `focus-metavariable: $X` is not a pattern in itself, it does not perform any matching, it only focuses the matching on the code already bound to `$X` by other patterns. Whereas `pattern: $X` matches `$X` against your code (and in this context, `$X` only matches expressions)!
```

Reviewer: Just a note for the use of future tense: https://www.notion.so/r2cdev/r2c-technical-communication-style-guide-4d1104e78927460aa8c7624aedd6561f#ca0bfefb2cfb42339cacc8eeb8a3554f
